rust/sbd-client/src/send_buf.rs

```diff
@@ -0,0 +1,159 @@
+use super::*;
+
+use std::collections::VecDeque;
+
+pub struct SendBuf {
+    pub ws: raw_client::WsRawSend,
+    pub buf: VecDeque<Vec<u8>>,
+    pub out_buffer_size: usize,
+    pub origin: tokio::time::Instant,
+    pub limit_rate: u64,
+    pub idle_keepalive_nanos: u64,
+    pub next_send_at: u64,
+    pub last_send: u64,
+}
+
+impl SendBuf {
+    /// construct a new send buf
+    pub fn new(
+        ws: raw_client::WsRawSend,
+        out_buffer_size: usize,
+        limit_rate: u64,
+        idle_keepalive: std::time::Duration,
+        pre_sent_bytes: usize,
+    ) -> Self {
+        let mut this = Self {
+            ws,
+            buf: VecDeque::default(),
+            out_buffer_size,
+            origin: tokio::time::Instant::now(),
+            limit_rate,
+            idle_keepalive_nanos: idle_keepalive.as_nanos() as u64,
+            next_send_at: 0,
+            last_send: 0,
+        };
+
+        let now = this.origin.elapsed().as_nanos() as u64;
+
+        this.next_send_at = std::cmp::max(now, this.next_send_at)
+            + (pre_sent_bytes as u64 * this.limit_rate);
+
+        this
+    }
+
+    /// Close the connection.
+    pub async fn close(&mut self) {
+        self.ws.close().await;
+    }
+
+    /// We received a new rate limit from the server, update our records.
+    pub fn new_rate_limit(&mut self, limit: u64) {
+        if limit < self.limit_rate {
+            // rate limit updates are sent on a best effort,
+            // and there are network timing conditions to worry about.
+            // assume we accidentally sent a message while the new limit
+            // was in effect, and accout for that in a brute-force manner.
+
+            let now = self.origin.elapsed().as_nanos() as u64;
+
+            self.next_send_at = std::cmp::max(now, self.next_send_at)
+                + (MAX_MSG_SIZE as u64 * self.limit_rate);
+        }
+        self.limit_rate = limit;
+    }
+
+    /// If we need to wait before taking the next step, this
+    /// returns how long.
+    pub fn next_step_dur(&self) -> Option<std::time::Duration> {
+        let now = self.origin.elapsed().as_nanos() as u64;
+
+        if now - self.last_send >= self.idle_keepalive_nanos {
+            // we need a keepalive now, don't wait
+            return None;
+        }
+
+        if now < self.next_send_at {
+            let need_keepalive_at =
+                self.idle_keepalive_nanos - (now - self.last_send);
+            let nanos =
+                std::cmp::min(need_keepalive_at, self.next_send_at - now);
```

Comment:
AFAIU this indicates a duration rather than an absolute point

```suggestion
            let need_keepalive_in =
                self.idle_keepalive_nanos - (now - self.last_send);
            let nanos =
                std::cmp::min(need_keepalive_in, self.next_send_at - now);
```